### PR TITLE
feat(manager): convert to typescript

### PR DESCRIPTION
**Pull request by viceice**
This pr coverts the whole manager part to typescript.

depends on #4155 :exclamation:

### CODE REVIEW

File: lib/manager/gradle/build-gradle.ts [hunk 1]
```diff
@@ -1,13 +1,34 @@
...
 /**
  * Functions adapted/ported from https://github.com/patrikerdes/gradle-use-latest-versions-plugin
  * gradle-use-latest-versions-plugin is licensed under MIT and Copyright (c) 2018 Patrik Erdes
  */
 
-let variables = {};
+let variables: Record<string, string> = {};
+
+// TODO: Unify with BuildDependency ?
+export interface GraddleDependency {
```
Comment (ikesyo) [inline lib/manager/gradle/build-gradle.ts]: typo: `Graddle` -> `Gradle`

File: lib/manager/gradle/gradle-updates-report.ts [hunk 2]
```diff
@@ -1,10 +1,33 @@
...
 const GRADLE_DEPENDENCY_REPORT_FILENAME = 'gradle-renovate-report.json';
 
-async function createRenovateGradlePlugin(localDir) {
+interface GraddleProject {
+  project: string;
+  repositories: string[];
+  dependencies: GraddleModule[];
+}
+
+interface GraddleModule {
```
Comment (ikesyo) [inline lib/manager/gradle/gradle-updates-report.ts]: Ditto.

File: lib/manager/gradle/gradle-updates-report.ts [hunk 3]
```diff
@@ -1,10 +1,33 @@
...
-const fs = require('fs-extra');
-const { logger } = require('../../logger');
+import { join } from 'path';
+import { writeFile, exists, readFile } from 'fs-extra';
+import { logger } from '../../logger';
 
 const GRADLE_DEPENDENCY_REPORT_FILENAME = 'gradle-renovate-report.json';
 
-async function createRenovateGradlePlugin(localDir) {
+interface GraddleProject {
```
Comment (ikesyo) [inline lib/manager/gradle/gradle-updates-report.ts]: Another `Graddle` here.

File: lib/manager/gradle/gradle-updates-report.ts [hunk 4]
```diff
@@ -1,10 +1,33 @@
...
+
+interface GraddleModule {
+  name: string;
+  group: string;
+  version: string;
+}
+
+type GraddleModuleWithRepos = GraddleModule & { repos: string[] };
+
+// TODO: Unify with GraddleDependency ?
```
Comment (ikesyo) [inline lib/manager/gradle/gradle-updates-report.ts]: Ditto.

**Comment by ViceIce** [comment]
Done, ready for review 🎉

File: lib/manager/circleci/update.ts [hunk 5]
```diff
@@ -1,11 +1,11 @@
...
+import { Upgrade } from '../common';
 
-module.exports = {
-  updateDependency,
-};
-
-function updateDependency(fileContent, upgrade) {
+export function updateDependency(
+  fileContent: string,
+  upgrade: Upgrade<{ lineNumber: number }>
```
Comment (rarkins) [inline lib/manager/circleci/update.ts]: Are these additions necessary now that we moved lineNumber into managerData?

File: lib/manager/common.ts [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+  internalPackages?: string[];
+  compatibility?: Record<string, string>;
+  datasource?: string;
+  registryUrls?: (string | Registry)[];
+  deps: PackageDependency[];
+  ignoreNpmrcFile?: boolean;
+  lernaClient?: string;
+  lernaPackages?: string[];
+  manager?: string;
+  mavenProps?: Record<string, any>;
```
Comment (rarkins) [inline lib/manager/common.ts]: Is there any automated way to determine if fields like this are. no longer needed?

**Review by rarkins — COMMENTED** [review]
I'm only 36 files in but wanted to submit the early review comments

File: lib/manager/circleci/update.ts (hunk 5 above)
Comment (ViceIce) [reply lib/manager/circleci/update.ts]: no, the generic part is optional, just to help type check. i can remove that.

File: lib/manager/common.ts (hunk 6 above)
Comment (ViceIce) [reply lib/manager/common.ts]: no, but vscode will show the reference count, if you use [codelense](https://code.visualstudio.com/blogs/2017/02/12/code-lens-roundup)

image

you can click on it to show where it is used.

File: lib/manager/circleci/update.ts (hunk 5 above)
Comment (ViceIce) [reply lib/manager/circleci/update.ts]: should i remove that?

Comment (rarkins) [reply lib/manager/circleci/update.ts]: I think for this one it's not necessary because lineNumber got moved to managerData

**Pull request merged**